# LeadNPC
## Systems Biology-Based Analysis Indicates Global Transcriptional Impairment in Lead-Treated Human Neural Progenitor Cells

This repository contain the files necessary to reproduce the results reported in our analysis of lead-treated human neural progenitor cells. The *bin* folder contains the code used to generate analysis and figures. It's based on R software (version 3.5.1) and uses the libraries biomaRt, affy, biomaRt, data.table, dplyr, edgeR, factoextra, FactoMineR, ggplot2, ggrepel, grid, gridExtra, hgu95av2.db, igraph, limma, purrr, RColorBrewer, RCy3, readxl, RedeR, scales, stats, stringr, tibble, tidyr, topGO, and transcriptogramer.

The *Data*  folder contains intermediate data files generated by the pipeline. So you can start the pipeline from scratch, downloading the raw data files from the experiment, or run just one phase using these files. Be sure to correct the path to data in the R scripts.

### Running the entire pipeline from scratch
In order, to run this pipeline you will need to download the  raw data sequence reads from Sequence Read Archive, accession SRP079342, Gene Expression Omnibus, [accession GSE84712](http://https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE84712 "accession GSE84712"), get the [Ensembl GRCh38 Human genome reference](ftp://ftp.ensembl.org/pub/release-91/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz) and [annotation](ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/Homo_sapiens.GRCh38.91.gtf.gz) (release 91), and have the software [Hisat2](http://ccb.jhu.edu/software/hisat2/dl/hisat2-2.1.0-Linux_x86_64.zip) and [FeatureCounts](https://sourceforge.net/projects/subread/files/subread-1.6.3/subread-1.6.3-Linux-x86_64.tar.gz) installed.

##### Create hisat2 index
> hisat2-build -p 8 ${genRefDir}"/Homo_sapiens.GRCh38.dna.primary_assembly.fa" ${genIndexDir}

##### Extract the splice sites
> extract_splice_sites.py ${genRefDir}"/Homo_sapiens.GRCh38.94.gtf" > ${genRefDir}"/Homo_sapiens.GRCh38.94.txt"

##### Generate the aligns
For each sample file do:
> hisat2 -x ${genIndexDir} --known-splicesite-infile ${genRefDir}"/Homo_sapiens.GRCh38.94.txt" -p 8 -1 ${sampleDir}/${file}_1".fastq.gz" -2 ${sampleDir}/${file}_2".fastq.gz"| samtools view -bS - > ${resultDir}/${file}".hisat.bam"; 

##### Count aligned reads 
First, generate the bam files list in a single line
> ls *.hisat.bam | tr '\n' ' '> bamList.txt

Then generate the count's file
> featureCounts -T 32  -t gene -g gene_id -a ${genRefDir}"/Homo_sapiens.GRCh38.94.gtf" -o ./allCountsHisat.txt $(cat bamList.txt)

Now you can use the R scripts as follow:
##### Create logCPM file
> 00ProcessCounts.R

##### PCA analysis and create transcriptogramer objects
> 1analisePCA.R

##### Plot transcriptogramer graphics
> 2plotTrancript.R

##### Cluster superposition analysis
> 4intersecClusters.R

##### Create clusters graphos
This is not a completely automatic process. You will need to use Cytoscape manually...
> 5graphTemposManual.R

##### Generate Figure 2 components
> 6BarraTempos.R

##### Generate Figures 3 and 4
This is not a completely automatic process. You will need to use RedeR and Cytoscape manually...
> 8graphoClustersRCy3.R

##### Generate Markers figures
> 14Marquers.R

##### Other analysis
Another auxiliary analysis was performed using several scripts placed inside /bin folder.
